libraries/chain/webassembly/runtimes/eos-vm-oc.cpp

```diff
@@ -62,12 +77,31 @@ std::unique_ptr<wasm_instantiated_module_interface> eosvmoc_runtime::instantiate
    return std::make_unique<eosvmoc_instantiated_module>(code_hash, vm_type, *this);
 }
 
+uint32_t eosvmoc_runtime::acquire_main_thread_exec_mem_index() {
+   return main_thread_index++;
+}
+
+void eosvmoc_runtime::release_main_thread_exec_mem_index() {
+   --main_thread_index;
+}
+
+uint32_t eosvmoc_runtime::acquire_ro_thread_exec_mem_index() {
+   return ro_thread_index++;
+}
+
+void eosvmoc_runtime::release_ro_thread_exec_mem_index() {
+   --ro_thread_index;
+}
+
 void eosvmoc_runtime::init_thread_local_data() {
-   exec_thread_local = std::make_unique<eosvmoc::executor>(cc);
-   mem_thread_local  = std::make_unique<eosvmoc::memory>(eosvmoc::memory::sliced_pages_for_ro_thread);
+   for (auto i = 0; i < 4; ++i) {
+      exec_thread_local.emplace_back(std::make_unique<eosvmoc::executor>(cc));
+      mem_thread_local.emplace_back(std::make_unique<eosvmoc::memory>(eosvmoc::memory::sliced_pages_for_ro_thread));
+   }
 }
 
-thread_local std::unique_ptr<eosvmoc::executor> eosvmoc_runtime::exec_thread_local{};
-thread_local std::unique_ptr<eosvmoc::memory> eosvmoc_runtime::mem_thread_local{};
+thread_local std::vector<std::unique_ptr<eosvmoc::executor>> eosvmoc_runtime::exec_thread_local{};
+thread_local std::vector<std::unique_ptr<eosvmoc::memory>> eosvmoc_runtime::mem_thread_local{};
+thread_local uint32_t eosvmoc_runtime::ro_thread_index = 0;
```

Comment:
I think we need the final stack approach before considering ready for review

Reply:
Done